### PR TITLE
(FFM-5171) horizontal scaling streams

### PR DESCRIPTION
**Pull request by conormurray95**
**Changes**
Enables horizontal scaling for proxies because they won't all publish messages to the same redis stream 
Enables streaming mode to work for in memory proxy because we don't rely on redis streams anymore

**Old flow**
Event enters sdk -> sdk passes it to us -> we put it on redis stream -> we read it off redis stream -> we pass it to pushpin -> pushpin sends to downstream sdks

**New flow**
Event enters sdk -> sdk passes it to us  -> we pass it to pushpin -> pushpin sends to downstream sdks

**Additional work**
- [x] Validate behaviour/error handling with direct linux/windows builds
- [x] Improve error handling
- [x] E2E tests - trigger changes on SaaS and verify we receive them for redis and in memory modes

### CODE REVIEW

File: cmd/ff-proxy/main.go [hunk 1]
```diff
@@ -359,6 +361,23 @@ func main() {
 
 		logger.Info("retrieved offline config")
 	} else {
+		gpc := gripcontrol.NewGripPubControl([]map[string]interface{}{
+			{
+				"control_uri": "http://localhost:5561",
+			},
+		})
+
+		var eventListener sdkStream.EventStreamListener
```
Comment (jcox250) [inline cmd/ff-proxy/main.go:370]: I see on [line 417](https://github.com/harness/ff-proxy/pull/99/files#diff-4ffaa8cb395a69efc0cd1ee9a7c1459387c4b3d3731d2519c82a27f640c671b9R417) this gets passed to the internal SDK startup code. 

If the stream healthcheck fails I think this eventListener will be nil? Just wondering if that could cause any issues with the internal SDK and if we should maybe implement a NoOp EventListener that we assign to this if the healthcheck fails?

File: stream/stream_worker.go [hunk 2]
```diff
@@ -19,116 +22,45 @@ type GripStream interface {
...
-				return err
-			}
-			s.log.Debug("succesfully published event to pushpin", "topic", topic, "channel", e.channel, "content", e.content)
-		}
+	if err := s.publish(ctx, streamEvent{
+		channel: topic,
+		content: content,
+		err:     nil,
+	}); err != nil {
+		s.log.Error("stream worker failed", "topic", topic)
```
Comment (jcox250) [inline stream/stream_worker.go:60]: Just wondering, should we be returning this error?

File: cmd/ff-proxy/main.go (hunk 1 above)
Comment (conormurray95) [reply cmd/ff-proxy/main.go:370]: Yes I'd dug into that myself, the go sdk does only forward events to that `Pub` function if the eventListener isn't nil so it should already be a no op if it's not set from what I can see 
https://github.com/harness/ff-golang-server-sdk/blob/main/stream/sse.go#L169

File: stream/stream_worker.go (hunk 2 above)
Comment (conormurray95) [reply stream/stream_worker.go:60]: For now I didn't and just log it here. If we return an error the go sdk will just log it again so wasn't sure it was helpful - https://github.com/harness/ff-golang-server-sdk/blob/main/stream/sse.go#L177

File: cmd/ff-proxy/main.go (hunk 1 above)
Comment (jcox250) [reply cmd/ff-proxy/main.go:370]: Ahh cool, should be sweet then

**Review by jcox250 — APPROVED** [review]

**Pull request merged**